File: generated_code/python/solution_98_1.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Determines if a binary tree is a valid Binary Search Tree (BST).
        
        Algorithm: Recursive Depth-First Search (DFS) with range constraints.
        Time Complexity: O(N) where N is the number of nodes, as we visit each node once.
        Space Complexity: O(H) where H is the height of the tree, representing the maximum 
                          depth of the recursion call stack (O(N) worst case, O(log N) best case).
        """
        def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
            # An empty tree/node is valid
            if not node:
                return True
            
            # The current node's value must sit strictly between the low and high boundaries
            if not (low < node.val < high):
                return False
            
            # Recursively validate subtrees:
            # - Left subtree values must be strictly less than the current node's value (updates high boundary)
            # - Right subtree values must be strictly greater than the current node's value (updates low boundary)
            return validate(node.left, low, node.val) and validate(node.right, node.val, high)
        
        # Initialize boundaries with -infinity and +infinity to handle standard integer limits
        return validate(root, float('-inf'), float('inf'))

File: generated_code/cpp/solution_98_1.cpp
#include <climits>

// Definition for a binary tree node.
// Included to make the code fully self-contained and compilable.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // We use long long tracking boundaries to prevent integer overflow issues 
        // with INT_MIN and INT_MAX node values.
        return validate(root, LLONG_MIN, LLONG_MAX);
    }

private:
    // Helper function that leverages a top-down DFS approach to pass down 
    // valid open-ended intervals for each node.
    // Time Complexity: O(N) where N is the number of nodes, as we visit each node once.
    // Space Complexity: O(H) where H is the height of the tree, due to the recursion stack.
    bool validate(TreeNode* node, long long low, long long high) {
        // An empty tree/sub-tree is inherently a valid BST.
        if (node == nullptr) {
            return true;
        }

        // The current node's value must strictly fall within the range (low, high).
        if (node->val <= low || node->val >= high) {
            return false;
        }

        // Recursively validate left and right subtrees:
        // For the left child, the upper limit updates to the current node's value.
        // For the right child, the lower limit updates to the current node's value.
        return validate(node->left, low, node->val) && validate(node->right, node->val, high);
    }
};